WebSocket based debugger interface

unknownbrackets opened the pull request:
If you want to try it:

1. Run this build on your device (Android / iOS / desktop / whatever.)
2. Enable Settings -> Tools -> Developer tools -> Allow remote debugger.
3. Go here: http://ppsspp-debugger.unknownbrackets.org/

Alternatively (only logs, no debugger):
Copy this into an HTML file (desktop, whatever): https://gist.github.com/unknownbrackets/385c01ca12a5ac22071f688e84c39ba3 (possibly PPSSPP could embed its own default HTML...)
Open the HTML file in a web browser and play a game: watch the logs roll in.

This is just a thrown-together example and of course command line, lua, whatever could be used to speak to this WebSocket ([a list of WebSocket libraries](https://github.com/facundofarias/awesome-websockets).)  I'm planning a simple JSON API.

The debugger would also work fine in a Chrome tab on the device (assuming a mobile friendly debugger UI...) as long as PPSSPP is not quit (which might become the case on iOS...)

 - [x] Update json #10960
 - [x] Fix debugger threading #10987
 - [x] IPv6 support #11004
 - [x] Include armips everywhere #11093
 - [x] Breakpoints on ARM #11148
 - [x] Remove Qt debugger #11130
 - [x] WebSocket / separated webserver #11152
 - [x] Rebase this (possibly merge some small commits together after rebase)

-[Unknown]

unknownbrackets committed:
Debugger: Serve a simple log listener WebSocket.

Planning to add more functionality to it, and sharing the reporting port
is possibly not ideal - although it would make discovery easier.

hrydgard commented on UI/RemoteISOScreen.cpp:
```diff
@@ -161,6 +162,10 @@ static void ExecuteServer() {
 	}
 
+	// TODO: This isn't an ideal place to put this... should separate.
+	// But kinda nice to only have it listen on one port?
```
Yeah, I think this is fine really.

unknownbrackets replied:
Cool.  Thinking of moving it to Core/WebServer.cpp or something, and sending flags to it to restart with remote iso / debugger enabled.

-[Unknown]

hrydgard commented:
Really cool stuff!

Regarding JSON serializer, we have a simple writer in ext/native/json_writer and a (different) reader in ext/native/ext/vjson. Might be worth using a proper one if we're gonna write a lot of different structs to json though, but I don't know much about which the best one today might be.

Same thing with React etc, not a frontend guy :)

unknownbrackets committed:
Debugger: Organize WebSocket event handling.

Just a starting point to organize it.  Trying to keep it simple.

unknownbrackets commented:
Okay, just using that json_writer then (totally forgot it existed.)  I think it will work fine.

Not sure if anyone else has opinions on frontend or where to source them - ideally this would make hacking, translation, cheat, whatever tools more accessible.

-[Unknown]

unknownbrackets committed:
Debugger: Synchronize error events using tickets.

unknownbrackets committed:
Debugger: Minor cleanup in MIPSDebugInterface.

unknownbrackets committed:
Debugger: Initial register get/set APIs.

unknownbrackets committed:
Debugger: Improve JSON number parsing.

With gason, this becomes doable without overflow issues.

unknownbrackets committed:
Debugger: Allow setting regs to inf/nan.

unknownbrackets committed:
Debugger: Cleanup, add initial API docs.

unknownbrackets commented:
Went ahead and switched from vjson to gason (newer from same author.)  Otherwise, vjson would parse 32-bit unsigned integer parameters as either truncated ints or floats.

-[Unknown]

hrydgard commented:
Yeah, gason seems fine. Good stuff.

unknownbrackets committed:
Debugger: Reorganize state handling.

Looking like this will be needed for disasm caches, etc.

unknownbrackets committed:
Debugger: Add cpu.stepping and cpu.resume.

unknownbrackets committed:
Debugger: Name the WebSocket debugger thread.

weituotian commented:
Really cool stuff!

if it is allow, i want `Memory::base`;╰(￣▽￣)╮

unknownbrackets commented:
I'm planning to have memory APIs: read, write, disasm, assemble, and memory breakpoint management.

Are you asking for an API to return the C++ address in Memory::base?  Guess I can add that (although it'll only be useful on some platforms, I think.)

-[Unknown]

unknownbrackets committed:
Debugger: Allow pinging current status.

Helps especially when reconnecting.

unknownbrackets committed:
Debugger: Use a counter when entering stepping.

This allows us to poll coreState and know if a new stepping needs to be
sent.  This is useful to allow e.g. regs to show changes per step.

unknownbrackets committed:
Debugger: Add config for remote debugger on start.

unknownbrackets committed:
Debugger: Add checkbox to enable remote debugger.

It's not a setting though, so maybe a button would be better.  Or, it
could be the startup setting...

unknownbrackets committed:
Debugger: Disconnect on shutdown/disable.

Although, it could be made safe to keep them up when restarting with
debugging still enabled.

unknownbrackets committed:
Log: Use a separate field for the timestamp.

Better than parsing it since we already have the header separate.  Simpler
too.

unknownbrackets committed:
Debugger: Add current game info.

unknownbrackets committed:
Debugger: Include PC and ticks in stepping events.

unknownbrackets committed:
Debugger: Add version event for future proofing.

Game isn't precisely right, but it's close enough.

unknownbrackets committed:
Debugger: Lock during startup/shutdown.

Otherwise things can get freed while we're trying to inspect them.

unknownbrackets committed:
Debugger: Initial disasm API.

unknownbrackets committed:
Debugger: Add additional encoding/symbol info.

unknownbrackets committed:
Debugger: Prevent crazy disasm range.

Also add some documentation.

unknownbrackets committed:
Debugger: Add cpu.evaluate method.

Probably useful for a bunch of things, really.

unknownbrackets committed:
Debugger: Add stepping to WebSocket API.

unknownbrackets commented:
Just a note: at this point the branch supports other platforms (Linux, Android, etc.) but there's no breakpoint support on ARM.  That means you can only step into and step HLE on ARM devices at the moment.

It also exposes Memory.base for all your cheat searching needs.

-[Unknown]

unknownbrackets commented:
Okay, I've published a somewhat functional debugger UI here:

http://ppsspp-debugger.unknownbrackets.org/

As noted above, x86 only for any breakpointing / step over / step out / etc.  You'll need to compile the latest branch here.

-[Unknown]

unknownbrackets committed:
Debugger: Use fragments for long chunks.

Gets it on the wire faster.  More importantly, this allows us to usually
avoid rampant memory allocation even for large responses.

unknownbrackets committed:
json: Optimize writing a bit.

This improves the responsiveness of certain APIs.

unknownbrackets committed:
Debugger: Poll frequently after stepping.

This will make us listen for events slightly less often, so we don't want
to sustain it.

unknownbrackets committed:
Debugger: Include assembler API.

unknownbrackets committed:
Debugger: Fix resume not skipping breakpoints.

unknownbrackets committed:
Debugger: Add thread safety to breakpoints.

unknownbrackets commented:
Now arm64 has breakpoints.  Didn't test memory breakpoints yet, but hopefully they work....

-[Unknown]

unknownbrackets committed:
Debugger: Avoid asserts in disassembly.

unknownbrackets committed:
Debugger: Disasm search API.

unknownbrackets committed:
Debugger: Initial breakpoint APIs.

unknownbrackets committed:
Debugger: Add memory breakpoint management.

unknownbrackets committed:
Debugger: Add some missing memory locks.

We can restart memory when loading save states, so we need this even while
we've got startup/shutdown locked.

unknownbrackets committed:
Debugger: Use a lock for memory reallocs.

Simpler this way, no need to remember to lock memory.

unknownbrackets committed:
Debugger: Add HLE API funcs.

Lumping function symbols in here too, they're pretty related...

unknownbrackets committed:
Debugger: Add debug interfaces for threads.

This way we can switch context.

unknownbrackets committed:
Debugger: Avoid some lock ordering issues.

Ideally get rid of the memory lock...

unknownbrackets committed:
Debugger: Allow stepping based on thread.

unknownbrackets commented:
New debugger (the multiplatform web one) now supports thread context debugging.  Put simply, you can now view and change thread regs, or select a thread and click "step into" to (mostly) step into that thread.

It's still based on address breakpoints (no thread condition) so it could break to a different thread.  Still a good improvement.

-[Unknown]

unknownbrackets committed:
Debugger: Allow conditions on threadID/moduleID.

And now step over/out/into can tie to the correct thread.

unknownbrackets committed:
Debugger: Add func name and data symbol to disasm.

unknownbrackets committed:
Debugger: Ignore invalid branches.

These happen on bytes that are not actually code.

unknownbrackets commented:
@hrydgard just a request in case you start merging any of the 1.7.x pulls next week: would be great not to merge them too quickly so the buildbot builds each merge.

Also not this pull yet, I want to rebase it (hence WIP.)

-[Unknown]

hrydgard commented:
Yeah, I know what we use WIP in titles for :)

I'll probably start merging towards the end of next week, hopefully 1.6.3 will be the last in the series.

unknownbrackets commented:
Heh, sorry.  I think it's good now - combined a few typo fixes into earlier commits.

At this point it's mostly usable as a debugger, although there's no memory view/search yet and nothing at all for GE debugger yet.

-[Unknown]

hrydgard commented:
Cool. I'll try to take a look soon - it's kind of a big one :)

unknownbrackets commented:
Heh, it's a couple few lines.  Happy to chop it up more if it helps, would be mostly linear at this point.

-[Unknown]

hrydgard commented:
Might be neat to pull the http: commits into a separate pull request which we'll merge first, but I'm not gonna demand anything :)

unknownbrackets commented:
Sure, separated out - #11152.

-[Unknown]

hrydgard commented:
please rebase on master so we get the buildfix that was just committed in any future bisect :)

unknownbrackets commented:
Yep, rebased.  I haven't tested these changes with GCC 8 either, though.

-[Unknown]

hrydgard commented:
Well, I guess there will be enough platforms that we can bisect on either way :) 

I'll give this a quick read through a little later and then probably merge.

hrydgard commented:
@unknownbrackets I'm fine with this. Again, it says "Not ready for merge" in the pull request, so should I or not? :)   If you're ready you can go ahead.

unknownbrackets commented:
Oops, thanks for checking - forgot to remove it from there.

-[Unknown]

leopard20 commented:
@unknownbrackets Finally! Let's check this out! :)

leopard20 commented:
@unknownbrackets So...how do I make this work on Android again? I enabled Allow Remote Debugging but when I go to the mentioned site it says it can't connect to PPSSPP.

unknownbrackets commented:
Hm.  That's all that should be needed.  Specifically:

1. Make sure both devices are connected to the same wifi and use NAT (e.g. you have local IPs like 10.x.x.x or 192.168.x.x or 172.16.x.x.)
2. Make sure Allow Remote Debugging is checked on the device you want to debug (e.g. Android.)
3. While PPSSPP is running (doesn't need to be in a game yet), open http://ppsspp-debugger.unknownbrackets.org/.

IMPORTANT: https://ppsspp-debugger.unknownbrackets.org/ will not work.  It unfortunately CANNOT be SSL because the emulator itself doesn't (and can't really) have a valid SSL certificate.

If this doesn't work, you can try looking at the IP and port your device is using (it can be seen under Tools -> Remote Disc Streaming -> Settings) and type it manually.  If you go to http://192.x.x.x:zzzz/ in your browser it should also show a text file with at least one line "/debugger".

If you can't access the ip:port from your PC browser, then it could mean something is blocking the connection (either not the same wifi, or your device has a firewall on it - most Androids don't.)  If you can, but the debugger doesn't work, it probably means a bug...

-[Unknown]

unknownbrackets commented:
Just wondering - did that help?

Also, I've outlined the missing features for the web debugger UI here:
https://github.com/unknownbrackets/ppsspp-debugger/projects

-[Unknown]

hrydgard commented:
Confirming that it works fine on Android for me if I have the device and the debugger on the same network. If it matters, whatismyip shows only a IPv4 address for me where I am now.

unknownbrackets commented:
Ah, yes, it might be a little restrictive for ipv6 currently.  It currently assumes the devices will be on the same /64 or "ipv6 subnet", which is commonly the case.

Your internet provider may give you an entire /56, but usually your router will assign IPs within a single /64 ipv6 subnet even so.  If the router ends up assigning IPs more liberally and your mobile device and desktop are in entirely different subnets, then the automatic IP detection won't work.

-[Unknown]

leopard20 commented:
Derp, forgot about my Android device firewall! Now it works! Thanks!
